### PR TITLE
Primitive voice implementation (feature/voice)

### PR DESCRIPTION
**Pull request by kozabrada123**
(opening this into a pr to track changes, progress, todos and merge conflicts)

**This pr (when complete) aims to add voice channel and call support to Chorus via Discord's UDP api.**

Current todo:

Features:
- [X] Voice gateway connection and heartbeating
- [X] Voice UDP connection
- [X] Ip Discovery
- [X] Select protocol
- [X] Receiving and deserializing audio
- [x] Audio decryption
- [x] Audio encryption
- [x] Sending and serializing audio
- [x] Public API

Docs:
- [x] Documentation on related structs, methods etc
- [x] Usage examples

Compatibility:
- [x] Gateway WASM combatibility
- [ ] ~~UDP WASM compatibility~~ - the browser **cannot** do udp.

Code quality:
- [x] Fix or at least clarify FIXMEs

## UDP vs Webrtc
User accounts sometimes use an undocumented Webrtc api for voice and video, while bots use the documented UDP api.
I've decided to implement the UDP api for now, since Webrtc is much more complex and less documented.
We may want to look into Webrtc aswell in the future.

## Architecture and public API
My current plan is to have four parts of the impl: the Voice Gateway, voice UDP, Voice Data and Voice Handler

- Voice Handler would handle coordinating and initializing the Gateway and UDP connections,
- Voice Data is a struct, which contains needed shared data, such as the encryption key, ssrc, session id, ip & port, etc..
- Voice Gateway and Voice UDP are the communication channels for Discord's api, using websockets and udp respectively.

The user can then interact with the Voice Gateway to issue gateway commands and receive voice context.
The voice UDP handler would expose a slightly higher level api, with listener events for decrypted and deserialized audio as well as a method which serializes, encrypts and sends audio.

### CODE REVIEW

**Comment by kozabrada123** [comment]
Update on this, I haven't had the time to work on this for *a while*, I'll see when I can restart this
the next thing I will probably try is encoding and decoding files with [a crate that might fix my previous issues]()https://crates.io/crates/symphonia

**Comment by kozabrada123** [comment]
If I can get decoding and encoding working in an example app, I can see whether the ddc client plays the audio. if not, then I've messed something up / something is still missing.

If it does play, I can ship the app's code as an example and merge this!

**Comment by kozabrada123** [comment]
The transport layer works perfectly, I just can't decode ddc's opus for some reason :/

File: src/types/events/voice_gateway/mod.rs [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+impl VoiceEncryptionMode {
+    /// Returns whether this encryption mode uses Xsalsa20Poly1305 encryption.
+    pub fn is_xsalsa20_poly1305(&self) -> bool {
+        match *self {
+            VoiceEncryptionMode::Xsalsa20Poly1305
+            | VoiceEncryptionMode::Xsalsa20Poly1305Lite
+            | VoiceEncryptionMode::Xsalsa20Poly1305Suffix
+            | VoiceEncryptionMode::Xsalsa20Poly1305LiteRtpsize => true,
+            _ => false,
+        }
```
[the remark this answers is not in the document]
Comment (kozabrada123) [reply src/types/events/voice_gateway/mod.rs:108]: Actually helpful lint, damn

File: src/voice/gateway/backends/tungstenite.rs [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+pub type TungsteniteSink =
+    SplitSink<WebSocketStream<MaybeTlsStream<TcpStream>>, tungstenite::Message>;
+pub type TungsteniteStream = SplitStream<WebSocketStream<MaybeTlsStream<TcpStream>>>;
+
+impl TungsteniteBackend {
+    pub async fn connect(
+        websocket_url: &str,
+    ) -> Result<(TungsteniteSink, TungsteniteStream), crate::errors::VoiceGatewayError> {
+        let mut roots = rustls::RootCertStore::empty();
+        for cert in rustls_native_certs::load_native_certs().expect("could not load platform certs")
```
Comment (bitfl0wer) [inline src/voice/gateway/backends/tungstenite.rs:29]: Can this loading of certs fail in any reasonable (especially outside-influenced) circumstance? If so, `.expect()` should be `?`. If `load_native_certs()` doesn't usually fail, `.expect()` is fine.

File: src/voice/gateway/gateway.rs [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+            url: websocket_url.clone(),
+            events: shared_events,
+            websocket_send: shared_websocket_send.clone(),
+            kill_send: kill_send.clone(),
+        })
+    }
+
+    /// The main gateway listener task;
+    ///
+    /// Can only be stopped by closing the websocket, cannot be made to listen for kill
```
Comment (bitfl0wer) [inline src/voice/gateway/gateway.rs:114]: Not that this is really relevant IMO, but can it really not be made to listen for kill? My naive instinct would be to pass a kill message through a channel

File: src/voice/gateway/handle.rs [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+        trace!("VGW: Requesting voice backend version");
+
+        self.send_json(VOICE_BACKEND_VERSION, data_empty_object)
+            .await;
+    }
+
+    /// Closes the websocket connection and stops all gateway tasks;
+    ///
+    /// Esentially pulls the plug on the voice gateway, leaving it possible to resume;
```
Comment (bitfl0wer) [inline src/voice/gateway/handle.rs:100]:
```suggestion
    /// Essentially pulls the plug on the voice gateway, leaving it possible to resume;
```

File: src/voice/gateway/message.rs [hunk 5]
```diff
@@ -0,0 +1,46 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+use crate::{errors::VoiceGatewayError, types::VoiceGatewayReceivePayload};
+
+/// Represents a messsage received from the voice websocket connection.
```
Comment (bitfl0wer) [inline src/voice/gateway/message.rs:7]:
```suggestion
/// Represents a message received from the voice websocket connection.
```

File: src/voice/gateway/message.rs [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+///
+/// This will be either a [VoiceGatewayReceivePayload], containing voice gateway events, or a [VoiceGatewayError].
+///
+/// This struct is used internally when handling messages.
+#[derive(Clone, Debug)]
+pub struct VoiceGatewayMessage(pub String);
+
+impl VoiceGatewayMessage {
+    /// Parses the message as an error;
+    /// Returns the error if succesfully parsed, None if the message isn't an error
```
Comment (bitfl0wer) [inline src/voice/gateway/message.rs:17]:
```suggestion
    /// Returns the error if successfully parsed, None if the message isn't an error
```

File: src/voice/udp/backends/tokio.rs [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+use crate::errors::VoiceUdpError;
+
+#[derive(Debug, Clone)]
+pub struct TokioBackend;
+
+pub type TokioSocket = tokio::net::UdpSocket;
+
+impl TokioBackend {
+    pub async fn connect(url: SocketAddr) -> Result<TokioSocket, VoiceUdpError> {
+        // Bind with a port number of 0, so the os assigns this listener a port
```
Comment (bitfl0wer) [inline src/voice/udp/backends/tokio.rs:16]: > So the os assigns this listener a port

Is this true only for Linux, or does this work on macOS and Windows as well?

File: src/voice/udp/handle.rs [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+                // "Incremental 4 bytes (32bit) int value"
+                let mut data_lock = self.data.write().await;
+                let nonce = data_lock
+                    .last_udp_encryption_nonce
+                    .unwrap_or_default()
+                    .wrapping_add(1);
+
+                data_lock.last_udp_encryption_nonce = Some(nonce);
+                drop(data_lock);
+                // TODO: Is le correct? This is not documented anywhere
```
Comment (bitfl0wer) [inline src/voice/udp/handle.rs:154]: Network byte order is usually BE. No clue about this specific case, though

File: src/voice/udp/handle.rs [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+            // return the error to the user.
+            //
+            // This is not an error the user should account for, which is why we throw it here.
+            panic!("{}", VoiceUdpError::FailedEncryption);
+        }
+
+        let mut encrypted_payload = encryption_result.unwrap();
+
+        // Append the nonce bytes, if needed
+        // All other encryption modes have an explicit nonce, where as Xsalsa20Poly1305
```
Comment (bitfl0wer) [inline src/voice/udp/handle.rs:215]:
```suggestion
        // All other encryption modes have an explicit nonce, whereas Xsalsa20Poly1305
```

File: src/voice/udp/handler.rs [hunk 10]
```diff
@@ -0,0 +1,356 @@
...
+#[derive(Debug)]
+/// The main UDP struct, which handles receiving, parsing and decrypting the rtp packets
+pub struct UdpHandler {
+    events: Arc<Mutex<VoiceUDPEvents>>,
+    pub data: Arc<RwLock<VoiceData>>,
+    socket: Arc<UdpSocket>,
+}
+
+impl UdpHandler {
+    /// Spawns a new udp handler and performs ip discovery.
```
Comment (bitfl0wer) [inline src/voice/udp/handler.rs:45]:
```suggestion
    /// Spawns a new udp handler and performs IP discovery.
```

File: src/voice/udp/handler.rs [hunk 11]
```diff
@@ -0,0 +1,356 @@
...
+    ///
+    /// Mutates the given data_reference with the ip discovery data.
```
Comment (bitfl0wer) [inline src/voice/udp/handler.rs:47]:
```suggestion
    /// Mutates the given data_reference with the IP discovery data.
```

File: src/voice/udp/handler.rs [hunk 12]
```diff
@@ -0,0 +1,356 @@
...
+    /// Receives udp messages and parses them.
+    async fn listen_task(&mut self) {
+        loop {
+            // FIXME: is there a max size for these packets?
+            // Allocating 512 bytes seems a bit extreme
+            //
+            // Update: see <https://stackoverflow.com/questions/58097580/rtp-packet-maximum-size>
+            // > "The RTP standard does not set a maximum size.."
+            //
+            // The theorhetical max for this buffer would be 1458 bytes, but that is imo
```
Comment (bitfl0wer) [inline src/voice/udp/handler.rs:153]:
```suggestion
            // The theoretical max for this buffer would be 1458 bytes, but that is imo
```

File: src/voice/voice_data.rs [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+/// Saves data shared between parts of the voice architecture;
+///
+/// Struct used to give the Udp connection data received from the gateway.
+pub struct VoiceData {
+    pub server_data: Option<VoiceServerUpdate>,
+    pub ready_data: Option<VoiceReady>,
+    pub session_description: Option<SessionDescription>,
+    pub user_id: Snowflake,
+    pub session_id: String,
+    /// The last sequence number we used, has to be incremeted by one every time we send a message
```
Comment (bitfl0wer) [inline src/voice/voice_data.rs:19]:
```suggestion
    /// The last sequence number we used, has to be incremented by one every time we send a message
```

File: examples/voice_simple/src/main.rs [hunk 14]
```diff
@@ -0,0 +1,311 @@
...
+        voice_data::VoiceData,
+    },
+};
+use log::{info, LevelFilter};
+use tokio::sync::{Mutex, RwLock};
+
+extern crate chorus;
+extern crate tokio;
+
+/// Handles inbetween connections between the gateway and udp modules
```
Comment (bitfl0wer) [inline examples/voice_simple/src/main.rs:43]:
```suggestion
/// Handles in between connections between the gateway and UDP modules
```

File: examples/voice_simple/src/main.rs [hunk 15]
```diff
@@ -0,0 +1,311 @@
...
+#[derive(Debug, Clone)]
+pub struct VoiceHandler {
+    pub voice_gateway_connection: Arc<Mutex<Option<VoiceGatewayHandle>>>,
+    pub voice_udp_connection: Arc<Mutex<Option<UdpHandle>>>,
+    pub data: Arc<RwLock<VoiceData>>,
+}
+
+impl VoiceHandler {
+    /// Creates a new voicehandler, only initializing the data
```
Comment (bitfl0wer) [inline examples/voice_simple/src/main.rs:52]:
```suggestion
    /// Creates a new [VoiceHandler], only initializing the data
```

File: examples/voice_simple/src/main.rs [hunk 16]
```diff
@@ -0,0 +1,311 @@
...
+}
+
+impl Default for VoiceHandler {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[async_trait]
+// On [VoiceServerUpdate] we get our starting data and url for the voice gateway server.
```
Comment (bitfl0wer) [inline examples/voice_simple/src/main.rs:69]:
```suggestion
// On [VoiceServerUpdate] we get our starting data and URL for the voice gateway server.
```

File: examples/voice_simple/src/main.rs [hunk 17]
```diff
@@ -0,0 +1,311 @@
...
+        voice_events
+            .ssrc_definition
+            .subscribe(self_reference.clone());
+
+        *self.voice_gateway_connection.lock().await = Some(voice_gateway_handle);
+    }
+}
+
+#[async_trait]
+// On [VoiceReady] we get info for establishing a UDP connection, and we immedietly need said UDP
```
Comment (bitfl0wer) [inline examples/voice_simple/src/main.rs:124]:
```suggestion
// On [VoiceReady] we get info for establishing a UDP connection, and we immediately need said UDP
```

File: src/errors.rs [hunk 18]
```diff
@@ -100,3 +100,59 @@ custom_error! {
...
+
+impl WebSocketEvent for VoiceGatewayError {}
+
+custom_error! {
+    /// Voice UDP errors.
+    #[derive(Clone, PartialEq, Eq)]
+    pub VoiceUdpError
+
+    // General errors
+    BrokenSocket{error: String} = "Could not write / read from udp socket: {error}",
```
Comment (bitfl0wer) [inline src/errors.rs:143]:
```suggestion
    BrokenSocket{error: String} = "Could not write / read from UDP socket: {error}",
```

File: src/errors.rs [hunk 19]
```diff
@@ -100,3 +100,59 @@ custom_error! {
...
+
+    // Encryption errors
+    EncryptionModeNotImplemented{encryption_mode: String} = "Voice encryption mode {encryption_mode} is not yet implemented.",
+    NoKey = "Tried to encrypt / decrypt rtp data, but no key has been received yet",
+    FailedEncryption = "Tried to encrypt rtp data, but failed. Most likely this is an issue chorus' nonce generation. Please open an issue on the chorus github: https://github.com/polyphony-chat/chorus/issues/new",
+    FailedDecryption = "Tried to decrypt rtp data, but failed. Most likely this is an issue chorus' nonce generation. Please open an issue on the chorus github: https://github.com/polyphony-chat/chorus/issues/new",
+    FailedNonceGeneration{error: String} = "Tried to generate nonce, but failed due to error: {error}.",
+
+    // Errors when initiating a socket connection
+    CannotBind{error: String} = "Cannot bind socket due to a udp error: {error}",
```
Comment (bitfl0wer) [inline src/errors.rs:154]:
```suggestion
    CannotBind{error: String} = "Cannot bind socket due to a UDP error: {error}",
```

File: src/types/events/voice_gateway/mod.rs [hunk 20]
```diff
@@ -0,0 +1,170 @@
...
+    pub op_code: u8,
+
+    #[serde(borrow)]
+    #[serde(rename = "d")]
+    pub data: &'a RawValue,
+}
+
+impl<'a> WebSocketEvent for VoiceGatewayReceivePayload<'a> {}
+
+/// The modes of encryption available in voice udp connections;
```
Comment (bitfl0wer) [inline src/types/events/voice_gateway/mod.rs:64]:
```suggestion
/// The modes of encryption available in voice UDP connections;
```

File: src/types/events/voice_gateway/ready.rs [hunk 21]
```diff
@@ -7,15 +7,27 @@ use std::net::Ipv4Addr;
...
 pub struct VoiceReady {
-    ssrc: i32,
-    ip: Ipv4Addr,
-    port: u32,
-    modes: Vec<String>,
+    /// See <https://developer.mozilla.org/en-US/docs/Web/API/RTCRtpStreamStats/ssrc>
+    pub ssrc: u32,
+    pub ip: Ipv4Addr,
+    pub port: u16,
+    /// The available encryption modes for the udp connection
```
Comment (bitfl0wer) [inline src/types/events/voice_gateway/ready.rs:26]:
```suggestion
    /// The available encryption modes for the UDP connection
```

File: src/types/events/voice_gateway/select_protocol.rs [hunk 22]
```diff
@@ -0,0 +1,52 @@
...
+/// The possible protocol for sending a receiving voice data.
+///
+/// See <https://discord-userdoccers.vercel.app/topics/voice-connections#select-protocol-structure>
+#[derive(Debug, Default, Serialize, Deserialize, Clone, Copy, PartialEq, Eq)]
+#[serde(rename_all = "snake_case")]
+pub enum VoiceProtocol {
+    #[default]
+    /// Sending data via UDP, documented and the only protocol chorus supports.
+    Udp,
+    // Possible value, yet NOT RECOMMENED, AS CHORUS DOES NOT SUPPORT WEBRTC
```
Comment (bitfl0wer) [inline src/types/events/voice_gateway/select_protocol.rs:37]:
```suggestion
    // Possible value, yet NOT RECOMMENDED, AS CHORUS DOES NOT SUPPORT WEBRTC
```

File: src/types/events/voice_gateway/select_protocol.rs [hunk 23]
```diff
@@ -0,0 +1,52 @@
...
+    //Webrtc,
+}
+
+#[derive(Debug, Default, Deserialize, Serialize, Clone)]
+/// The data field of the SelectProtocol Event
+///
+/// See <https://discord-userdoccers.vercel.app/topics/voice-connections#protocol-data-structure>
+pub struct SelectProtocolData {
+    /// Our external ip we got from ip discovery
```
Comment (bitfl0wer) [inline src/types/events/voice_gateway/select_protocol.rs:46]:
```suggestion
    /// Our external IP we got from IP discovery
```

File: src/types/events/voice_gateway/select_protocol.rs [hunk 24]
```diff
@@ -0,0 +1,52 @@
...
+    pub address: String,
+    /// Our external udp port we got from id discovery
```
Comment (bitfl0wer) [inline src/types/events/voice_gateway/select_protocol.rs:48]:
```suggestion
    /// Our external UDP port we got from IP discovery
```

File: src/types/events/voice_gateway/speaking.rs [hunk 25]
```diff
@@ -0,0 +1,52 @@
...
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+use bitflags::bitflags;
+use serde::{Deserialize, Serialize};
+
+use crate::types::{Snowflake, WebSocketEvent};
+
+/// Event that tells the server we are speaking;
+///
+/// Essentially, what allows us to send udp data and lights up the green circle around your avatar.
```
Comment (bitfl0wer) [inline src/types/events/voice_gateway/speaking.rs:12]:
```suggestion
/// Essentially, what allows us to send UDP data and lights up the green circle around your avatar.
```

File: src/voice/voice_data.rs [hunk 26]
```diff
@@ -0,0 +1,25 @@
...
+    pub last_sequence_number: u16,
+    pub ip_discovery: Option<IpDiscovery>,
+
+    /// The last udp encryption nonce, if we are using an encryption mode with incremental nonces.
```
Comment (bitfl0wer) [inline src/voice/voice_data.rs:23]:
```suggestion
    /// The last UDP encryption nonce, if we are using an encryption mode with incremental nonces.
```

File: src/voice/udp/handle.rs [hunk 27]
```diff
@@ -0,0 +1,256 @@
...
+
+use crate::{
+    errors::VoiceUdpError,
+    types::VoiceEncryptionMode,
+    voice::{crypto::get_xsalsa20_poly1305_nonce, voice_data::VoiceData},
+};
+
+use super::{events::VoiceUDPEvents, RTP_HEADER_SIZE};
+
+/// Handle to a voice udp connection
```
Comment (bitfl0wer) [inline src/voice/udp/handle.rs:27]:
```suggestion
/// Handle to a voice UDP connection
```

File: src/voice/udp/handle.rs [hunk 28]
```diff
@@ -0,0 +1,256 @@
...
+    ///
+    /// Automatically makes an [RtpPacket](discortp::rtp::RtpPacket), encrypts it and sends it.
+    ///
+    /// # Errors
+    /// If we do not have VoiceReady data, which contains our ssrc, this returns a
+    /// [VoiceUdpError::NoData] error.
+    ///
+    /// If we have not received an encryption key, this returns a [VoiceUdpError::NoKey] error.
+    ///
+    /// If the Udp socket is broken, this returns a [VoiceUdpError::BrokenSocket] error.
```
Comment (bitfl0wer) [inline src/voice/udp/handle.rs:48]:
```suggestion
    /// If the UDP socket is broken, this returns a [VoiceUdpError::BrokenSocket] error.
```

File: src/voice/udp/handler.rs [hunk 29]
```diff
@@ -0,0 +1,356 @@
...
+        Ok(UdpHandle {
+            events: shared_events,
+            socket,
+            data: data_reference,
+        })
+    }
+
+    /// The main listen task;
+    ///
+    /// Receives udp messages and parses them.
```
Comment (bitfl0wer) [inline src/voice/udp/handler.rs:144]:
```suggestion
    /// Receives UDP messages and parses them.
```

File: src/voice/udp/mod.rs [hunk 30]
```diff
@@ -0,0 +1,18 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+//! Defines the udp component of voice communications, sending and receiving raw rtp data.
```
Comment (bitfl0wer) [inline src/voice/udp/mod.rs:5]:
```suggestion
//! Defines the UDP component of voice communications, sending and receiving raw rtp data.
```

**Review by bitfl0wer — CHANGES_REQUESTED** [review]
Looks good to me! Amazing work :)
I'd only like the 1-2 questions I had and the spelling mistake thingies to be resolved to be merging, but otherwise this looks great! :D

File: src/voice/udp/backends/tokio.rs (hunk 7 above)
Comment (kozabrada123) [reply src/voice/udp/backends/tokio.rs:16]: It should be true regardless of OS, according to the [tokio docs](https://docs.rs/tokio/latest/tokio/net/struct.UdpSocket.html#method.bind)

File: src/voice/gateway/backends/tungstenite.rs [hunk 31]
```diff
@@ -0,0 +1,77 @@
...
+impl TungsteniteBackend {
+    pub async fn connect(
+        websocket_url: &str,
+    ) -> Result<(TungsteniteSink, TungsteniteStream), crate::errors::VoiceGatewayError> {
+        let mut roots = rustls::RootCertStore::empty();
+        let certs = rustls_native_certs::load_native_certs();
+
+        if let Err(e) = certs {
+            log::error!("Failed to load platform native certs! {:?}", e);
+            return Err(GatewayError::CannotConnect {
```
[the remark this answers is not in the document]
Comment (kozabrada123) [reply src/voice/gateway/backends/tungstenite.rs]: shush

File: src/voice/udp/handle.rs (hunk 8 above)
Comment (kozabrada123) [reply src/voice/udp/handle.rs]: True, I've changed it to big endian

File: src/voice/gateway/backends/tungstenite.rs (hunk 2 above)
Comment (kozabrada123) [reply src/voice/gateway/backends/tungstenite.rs]: It apparently can in some platform specific ways, I've added a quick check along with returning `VoiceGatewayError::CannotConnect`

File: src/voice/gateway/gateway.rs (hunk 3 above)
Comment (kozabrada123) [reply src/voice/gateway/gateway.rs:114]: Perhaps not that it strictly _can't_ listen for the existing kill channel, it's just a lot easier to use the closing of the websocket

The task runs in a loop, so it's almost always awaiting a new websocket message, such that when the websocket closes it halts the loop.

We could use tokio::select to await messages on the kill channel or the websocket, but that would make the code unnecessarily more complex imo

**Review by bitfl0wer — APPROVED** [review]

File: src/gateway/gateway.rs [hunk 32]
```diff
@@ -103,7 +105,17 @@ impl Gateway {
     /// Can only be stopped by closing the websocket, cannot be made to listen for kill
```
Comment (bitfl0wer) [inline src/gateway/gateway.rs:105]: remove this docstring :3

Comment (kozabrada123) [reply src/gateway/gateway.rs:105]: ah, forgor

**Pull request merged**